### ISSUE TITLE
Rolling restart of pods

### ISSUE THREAD

**Issue by ghodss** [kubernetes/kubernetes#13488]
`kubectl rolling-update` is useful for incrementally deploying a new replication controller. But if you have an existing replication controller and want to do a rolling restart of all the pods that it manages, you are forced to do a no-op update to an RC with a new name and the same spec. It would be useful to be able to do a rolling restart without needing to change the RC or to give the RC spec, so anyone with access to kubectl could easily initiate a restart without worrying about having the spec locally, making sure it's the same/up to date, etc. This could work in a few different ways:
1. A new command, `kubectl rolling-restart` that takes an RC name and incrementally deletes all the pods controlled by the RC and allows the RC to recreate them.
2. Same as 1, but instead of deleting each pod, the command iterates through the pods and issues some kind of "restart" command to each pod incrementally (does this exist? is this a pattern we prefer?). The advantage of this one is that the pods wouldn't get unnecessarily rebalanced to other machines.
3. `kubectl rolling-update` with a flag that lets you specify an old RC only, and it follows the logic of either 1 or 2.
4. `kubectl rolling-update` with a flag that lets you specify an old RC only, and it auto-generates a new RC based on the old one and proceeds with normal rolling update logic.

All of the above options would need the MaxSurge and MaxUnavailable options recently introduced (see #11942) along with readiness checks along the way to make sure that the restarting is done without taking down all the pods.

@nikhiljindal @kubernetes/kubectl

**Comment by nikhiljindal** [comment]
cc @ironcladlou @bgrant0607 

Whats the use case for restarting the pods without any changes to the spec?

Note that there wont be any way to rollback the change if pods started failing when they were restarted.

**Comment by ghodss** [comment]
Whenever services get into some wedged or undesirable state (maxed out connections and are now stalled, bad internal state, etc.). It's usually one of the first troubleshooting steps if the service is seriously misbehaving.

If the first pod fails as it is restarted, I would expect it to cease continuing or continue retrying to start the pod.

**Comment by smarterclayton** [comment]
Also, a rolling restart with no spec change reallocates pods across the
cluster.

However, I would also like the ability to do this without rescheduling the
pods.  That could be a rolling label change, but may pick up new dynamic
config or clear the local file state.

On Wed, Sep 2, 2015 at 12:01 AM, Sam Ghods <notifications@github.com> wrote:

> Whenever services get into some wedged or undesirable state (maxed out
> connections and are now stalled, bad internal state, etc.). It's usually
> one of the first troubleshooting steps if the service is seriously
> misbehaving.
>
> If the first pod fails as it is restarted, I would expect it to cease
> continuing or continue retrying to start the pod.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/issues/13488#issuecomment-136931790>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by ghodss** [comment]
@smarterclayton Is that like my option 2 listed above? Though why would labels be changed?

**Comment by bgrant0607** [comment]
Re. wedged: That's what liveness probes are for.

Re. rebalancing: see #12140

If we did support this, I'd lump it with #9043 -- the same mechanism is required.

**Comment by ghodss** [comment]
I suppose this would more be for a situation where the pod is alive and responding to checks but still needs to be restarted. One example is a service with an in-memory cache or internal state that gets corrupted and needs to be cleared.

I feel like asking for an application to be restarted is a fairly common use case, but maybe I'm incorrect.

**Comment by bgrant0607** [comment]
Corruption would just be one pod, which could just be killed and replaced by the RC.

The other case mentioned offline was to re-read configuration. That's dangerous to do implicitly, because restarts for any reason would cause containers to load the new configuration. It would be better to do a rolling update to push a new versioned config reference (e.g. in an env var) to the pods. This is similar to what motivated #1353.

**Comment by gmarek** [comment]
@bgrant0607 have we decided that we don't want to do this?

**Comment by bgrant0607** [comment]
@gmarek Nothing, for now. Too many things are underway already.

**Comment by gmarek** [comment]
Can we have a `post v1.1` milestone (or something) for the stuff that we deem important, but we lack people to fix them straight away?

**Comment by Glennvd** [comment]
I would be a fan of this feature as well, you don't want to be forced to switch tags for every minor update you want to roll out.

**Comment by mbmccoy** [comment]
I'm a fan of this feature. Use case: Easily upgrade all the pods to use a newly-pushed docker image (with `imagePullPolicy: Always`). I currently use a bit of a hacky solution: Rolling-update with or without the `:latest` tag on the image name.

**Comment by mbmccoy** [comment]
Another use case: Updating secrets.

**Comment by ericuldall** [comment]
I'd really like to see this feature. We run node apps on kubernetes and currently have certain use cases where we restart pods to clear in app pseudo caching.

Here's what I'm doing for now:

```
kubectl get pod | grep 'pod-name' | cut -d " " -f1 - | xargs -n1 -P 10 kubectl delete pod
```

This deletes pods 10 at a time and works well in a replication controller set up. It does not address any concerns like pod allocation or new pods failing to start. It's a quick solution when needed.

**Comment by jonaz** [comment]
I would really like to be able to do a rolling restart. 
The main reason is we will feed ENV variables into pods using ConfigMap and then if we change config we need to restart the consumers of that ConfigMap.

**Comment by paunin** [comment]
Yes, there are a lot of cases when you really want to restart pod/container without changes inside...
Configs, cache, reconnect to external services, etc. I really hope feature will be developed.

**Comment by paunin** [comment]
Small work around (I use deployments and I want to change configs without having real changes in image/pod):
* create configMap
* create deployment with ENV variable (you will use it as indicator for your deployment) in any container
* update configMap
* update deployment (change this ENV variable)

k8s will see that definition of deployment has been changed ad will start process of replacing pods
PS:
if someone have better solution please share

**Comment by Lasim** [comment]
Thank you @paunin

**Comment by w0mbat** [comment]
@paunin Thats exactly the case where we need it currently - We have to change ConfigMap values that are very important to the services and need to be rolled-out to the containers within minutes up to some hours. If no deployment happens in the meantime the containers will all fail at the same time and we will have partial downtime of at least some seconds

**Comment by rcoup** [comment]
from (kinda-related #9043): a one-liner of @paunin's approach, where `RESTART_` is the environment variable:
```
kubectl patch deployment mydeployment \
-p'{"spec":{"template":{"spec":{"containers":[{"name":"mycontainer","env":[{"name":"RESTART_","value":"$(date -uIseconds)"}]}]}}}}'
```

(note for some reason environment variables starting with `_` just seem to disappear, and numeric env `value`s cause errors, they need to be strings)

**Comment by ghodss** [comment]
@paunin @rcoup We do something very similar nowadays, we have an env var literally called "DUMMY_VAR_FOR_NO_OP_DEPLOYMENT".

It feels like the right solution here would enable you to restart a deployment, and reuse most of the deployment parameters for rollouts like MinReadyCount, while allowing for command-line overrides like increasing the parallelism for emergency situations where you need everything to bounce immediately.

**Comment by alphashuro** [comment]
any progress on this?

**Comment by kargakis** [comment]
There is a proposal and PoC in https://github.com/kubernetes/kubernetes/pull/31701

**Comment by gajus** [comment]
I feel like this addition is just an unnecessary bloating of the CLI API. This can be easily achieved by updating deployment environment variable values [as suggested by @paunin](https://github.com/kubernetes/kubernetes/issues/13488#issuecomment-218093109).

**Comment by montanaflynn** [comment]
We would also like to see this for deployments maybe like `kubectl restart deployment some-api`

**Comment by whereisaaron** [comment]
Kubernetes is allowed to restart Pods for all sorts of reasons, but the cluster admin isn't allowed to. 
I understand the moral stand that 'turn it off and on again' may not be a desired way to operate... but I also think it should be ok to let those people who wish to, to restart a Deployment without resorting to the range of less appetizing tricks like:
- deleting pods
- dummy labels
- dummy environment variables
- dummy config maps mapped to environment variable
- rebooting the worker nodes
- cutting the power to the data centre 😄 

'No, no, I'm not restarting anything, just correcting a typo in this label here' 😛

**Comment by DmitryRomanenko** [comment]
This feature will be useful in pair with `kubectl apply`: `apply` will update configs, including Replication Controllers, but pods won't be restarted.

So we need a method to restart these pods Blue-Green way.

**Comment by kargakis** [comment]
@DmitryRomanenko how about switching from ReplicationControllers to Deployments? That will enable you to run updates of ReplicaSets (successor to ReplicationController).

**Comment by DmitryRomanenko** [comment]
@kargakis it's not possible: Deployments update only Replica Sets and Pods.  
With `kubectl apply` we also update ConfigMaps, Services, etc.

**Comment by tyranron** [comment]
@DmitryRomanenko if the issue is "I want to restart Pods when ConfigMap/Secret is updated" the possible solution is to have versions for your ConfigMap and Secrets, and make that version a part of your Deployment spec. So with `kubectl apply`ing the spec of Deployment is changed and Pods are recreated.
In other situations I don't see why Pods must be restarted (I mean Service/Ingress/etc update).

**Comment by DmitryRomanenko** [comment]
@tyranron, thanks! What is the best way to version `ConfigMap`s? Should I create new config map with a different name for new deployment?

**Comment by tyranron** [comment]
@DmitryRomanenko you actually can, why not? But in this case you should take care about removing old ones. On the other hand it may be useful for rollback. But in most cases specifying version via labels is enough.

**Comment by paunin** [comment]
I do believe that the best solution here could be some kind of watcher or hash-sum checker on `configmap` object. Which should trigger related objects/pods restart(anything uses that `configmap`,`secret`). Not sure it's something accessible in `k8s` architecture though...

I also think that it's better to have control from `configmap|secret` object to trigger restart on changes or not.

**Comment by kargakis** [comment]
@paunin you may want to have a look into a custom controller I created as a poc a while ago. It watches configmaps and hashes cm content to create copies and mount them into deployments.
https://github.com/kargakis/configmap-rollout
If you are familiar with Go, I would suggest writting something similar in Go (you would get better error handling and access to all k8s libraries). Actually there is https://github.com/fabric8io/configmapcontroller but I haven't looked closely.

**Comment by DmitryRomanenko** [comment]
@tyranron,
> So with kubectl applying the spec of Deployment is changed and Pods are recreated.

Could you explain? Should I just use `kubectl apply -f new_config.yml` with updated deployments, and these deployments will be rolling-restarted?

**Comment by tyranron** [comment]
@DmitryRomanenko yup, exactly.

**Comment by tyranron** [comment]
@DmitryRomanenko with applying new specs you are [updating Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment#updating-a-deployment), and on Deployment update restart is triggered if its spec was changed.

By default the restart strategy is `RollingUpdate `, but you can specify another one explicitly too.

**Comment by vdboor** [comment]
A small change to #issuecomment-240393845: make sure the `date` is evaluated within the shell:

```
kubectl patch deployment mydeployment -p '{"spec":{"template":{"spec":{"containers":[{"name":"mycontainer","env":[{"name":"RESTART_","value":"'$(date +%s)'"}]}]}}}}'
```

**Comment by hammady** [comment]
Been using Swarm mode for some time which is considered less flexible than Kubernetes, I am able to restart service tasks (read: deployment pods), by simply doing a force update (with no change to the specs) as in `docker service update --force <service-name>`. I would expect Kubernetes to do this as well.
As for configmaps and secrets, swarm does not allow you to edit those, you need to rotate them instead. You do this by creating new configmaps/secrets, updating the service spec to use the new ones, then delete the older ones. I see this is typically what is recommended above by versioning your configmaps/secerts and updating the deployments that use them. To be honest, this behavior of rotation is one major reason I left swarm for! It is very inconvenient to have a local copy, update then create new resources and finally update the dependent resources. Add to that, secrets in swarm cannot be read from the api, you have to mount them inside any container (or exec inside a container that uses them), then `cat` their files.
On a related note, I used openshift for some time and I believe it auto-restarts pods on env/configmap/secret change? I stand corrected though.

**Comment by so0k** [comment]
should be app responsibility to watch filesystem for changes, as mentioned you can use checksums on the configmap/secret and force restarts that way

but if you don't want to change the config at all and just do a rolling restart with arbitrary pause, a simple pipeline does the job

```bash
kubectl get po -l release=my-app -o name | cut -d"/" -f2 | while read p;do kubectl  delete po $p;sleep 30;done
```

**Comment by KIVagant** [comment]
@so0k alternative command:
```kubectl get pods|grep somename|awk '{print $1}' | xargs -i sh -c 'kubectl delete pod -o name {} && sleep 4'
```

**Comment by whereisaaron** [comment]
Two and half years on and people are still crafting new workarounds, with dummy env vars, dumy labels, ConfigMap and Secret watcher sidecars, scaling to zero, and straight out rolling-update shell scripts to simulate the ability the trigger a rolling update. Is this still something cluster admins should not be allowed to do honestly, without the tricks?

#27081 #33664 https://github.com/kubernetes/helm/issues/2639

https://stackoverflow.com/questions/41735829/update-a-deployment-image-in-kubernetes
```
kubectl scale --replicas=0 deployment application
kubectl scale --replicas=1 deployment application
```

https://stackoverflow.com/questions/40366192/kubernetes-how-to-make-deployment-to-update-image
> Another trick is to intially run:
> ```
> kubectl set image deployment/my-deployment mycontainer=myimage:latest
> ```
> and then:
> ```
> kubectl set image deployment/my-deployment mycontainer=myimage
> ```
> It will actually be triggering the rolling-update but be sure you have also imagePullPolicy: "Always" set.
> 

> another trick I found, where you don't have to change the image name, is to change the value of a field that will trigger a rolling update, like terminationGracePeriodSeconds. You can do this using kubectl edit deployment your_deployment or kubectl apply -f your_deployment.yaml or using a patch like this:
> ```
> kubectl patch deployment your_deployment -p \
>   '{"spec":{"template":{"spec":{"terminationGracePeriodSeconds":31}}}}'
> ```
> 

http://rancher.com/docs/rancher/v1.4/en/cattle/upgrading/
```
# Force an upgrade even though the docker-compose.yml for the services didn't change
$ rancher-compose up --force-upgrade
```

**Comment by vdboor** [comment]
@so0k @KIVagant deleting pods means downtime, even when running multiple instances. When someone runs a single pod with `strategy.rollingUpdate.maxUnavailable = 0` kubernetes creates a new pod first before it terminates the existing one. The `kubectl patch deployment` trick triggers this behavior, and I'd really like to see this being a default one.

For instance, when running images from hub.docker.com, the same tag can be patched for security updates. I'd really like to "pull the latest images", and perform a rolling update for any outdated image.

**Comment by joelittlejohn** [comment]
@bgrant0607 I think the important use-case that is not covered by other tickets is this one: https://github.com/kubernetes/kubernetes/issues/13488#issuecomment-136946912

> @ghodss wrote:
> I suppose this would more be for a situation where the pod is alive and responding to checks but still needs to be restarted. One example is a service with an in-memory cache or internal state that gets corrupted and needs to be cleared.
> 
> I feel like asking for an application to be restarted is a fairly common use case, but maybe I'm incorrect.

I'd like to force a rolling restart to clear all application state without manual tricks.

**Comment by mattdodge** [comment]
I've got a similar one-liner based on the approach that @rcoup and @paunin described, but it should work for the more general case. It relies on JSON output and uses jq for parsing. I have this loaded in to my shell as a function.

```bash
kubectl-restart() {
  kubectl get deploy $1 -o json | jq \
    'del(
      .spec.template.spec.containers[0].env[]
      | select(.name == "RESTART_"))
    | .spec.template.spec.containers[0].env += [{name: "RESTART_", value: now|tostring}]' | kubectl apply -f -
}
```

This allows me to just say: ```kubectl-restart my-deployment-name``` and it will "update" the `RESTART_` env var in the first container to the current timestamp. I'm far from a jq expert so there may be a better way to do it, but basically it deletes the old `RESTART_` env var from the output (if it exists) and then adds it back in there with the current time.

It does feel very weird to me that there is no native way to do this though... Certainly a room full of engineers would agree that the ability to "turn it off and back on again" is something we'd like to have.

**Comment by gjcarneiro** [comment]
That is a fine hack and all, but it has big downside.  The next time you deploy, using `kubectl apply -f`, that component will get restarted if it has a RESTART_xxx env var even if nothing else changes.  In other words, it causes spurious restarts on next deployment if it has ever been restarted between last deployment and this deployment.  Not ideal...

That is why rolling-restart functionality should be baked into the Deployment controller, not built on top.

**Comment by jmound** [comment]
I wrote a bash function to accomplish the "patch deployment of `terminationGracePeriodSeconds`" strategy @whereisaaron quoted in his [comment](https://github.com/kubernetes/kubernetes/issues/13488#issuecomment-372532659) above (source: https://stackoverflow.com/a/40368520/90442).

```
# $1 is a valid namespace
function refresh-all-pods() {
  echo
  DEPLOYMENT_LIST=$(kubectl -n $1 get deployment -o json|jq -r .items[].metadata.name)
  green "Refreshing pods in all Deployments"
  for deployment_name in $DEPLOYMENT_LIST ; do
    TERMINATION_GRACE_PERIOD_SECONDS=$(kubectl -n $1 get deployment "$deployment_name" -o json|jq .spec.template.spec.terminationGracePeriodSeconds)
    if [ "$TERMINATION_GRACE_PERIOD_SECONDS" -eq 30 ]; then
      TERMINATION_GRACE_PERIOD_SECONDS='31'
    else
      TERMINATION_GRACE_PERIOD_SECONDS='30'
    fi
    patch_string="{\"spec\":{\"template\":{\"spec\":{\"terminationGracePeriodSeconds\":$TERMINATION_GRACE_PERIOD_SECONDS}}}}"
    kubectl -n $1 patch deployment $deployment_name -p $patch_string
  done
  echo
}
```

Comment / update it via the [gist here](https://gist.github.com/jmound/ff6fa539385d1a057c82fa9fa739492e). I'll echo other's that this is not ideal behavior.

**Comment by tcolgate** [comment]
Just by way of a more specifically kube related justification, restart also allows the re-running of an init-container, which can be used to roll keys, update config, or whatever you use init containers for.

**Comment by bgrant0607** [comment]
@kubernetes/sig-apps-feature-requests @kow3ns @janetkuo

**Comment by bgrant0607** [comment]
@gjcarneiro Did you have a RESTART_xxx env var in the configuration you applied, or not? If not, then I'd expect apply to ignore the extra env var in the live state.

cc @apelisse

**Comment by apelisse** [comment]
@gjcarneiro Yeah, the problem with @mattdodge script is that it's using apply, so the change will be saved in the lastApplied annotation. The script could be fixed by using patch or another method to update the deployment.

**Comment by sbrichardson** [comment]
This is ridiculous and needs to be fixed.

**Comment by brycesteinhoff** [comment]
Would love to have this feature. It seems very basic and needed.

**Comment by KIVagant** [comment]
@alcohol , check this one https://github.com/kubernetes/kubernetes/issues/13488#issuecomment-356892053

**Comment by alcohol** [comment]
Thanks for the tip :-)

**Comment by megakoresh** [comment]
Openshift has the concept of deployment triggers, which trigger a rollout on image change, webhook or configuration change. It would be very good feature to have in Kubernetes. As well as manual rollouts of course.

Furthermore Docker repository has history so there is no reason why rollback couldn't work - pod spawned from `.spec.template` can use `image-tag:@digest` format when pulling images for containers. Rollback would use the digest ID of the previous rollout.

**Comment by realfresh** [comment]
Not sure if I'm understanding correctly. Just in case this helps anybody. 

It seems that if you update the value of a label under the pod > template > metadata, then a rolling update takes place after you `kubectl apply -f file.yaml`

So you can always have a label for your version and whenever you want to rolling update, change the version and apply the file.

**Comment by joelittlejohn** [comment]
@gjcarneiro If you don't `apply` when you make a change, the `kubectl.kubernetes.io/last-applied-configuration` annotation will not get updated, so the next `apply` will not cause another restart.

I'm strongly in favour of adding a rolling restart command to kubectl, but in the meantime I'm using the following (based on solutions above):

```bash
kubectl patch deployment mydeployment -p '{"spec":{"template":{"spec":{"containers":[{"name":"mycontainer","env":[{"name":"RESTART_","value":"'$(date +%s)'"}]}]}}}}'
```

Parameterize this and add it as a function in your .bashrc and it's a good interim solution.

**Comment by gjcarneiro** [comment]
Ah, cool, I didn't know that, thanks!

I don't need the bash alias, at my company we made my own web interface for managing Kubernetes using Python+aiohttp and it  already uses patch.  I thought about open sourcing it, just been lazy...

**Comment by matti** [comment]
Looks like people are repeating the same workaround solutions in this thread - please read the full thread before posting here

**Comment by Macmee** [comment]
@joelittlejohn I ran your macro and it DID trigger a reboot of my pods, but they all restarted at the same time. I thought this would trigger a rolling reboot, no?

**Comment by joelittlejohn** [comment]
@Macmee It depends on the configuration of your deployment. The above command just changes the deployment. The pods are then updated according to the roll out `strategy` defined by your deployment. This is just like any other change to the deployment.

The only way this will replace all pods at the same time is if your `.spec.strategy.rollingUpdate.maxUnavailable` allows it.

**Comment by japzio** [comment]
one use-case also on our side is that we use spring-cloud-config-server for our spring-boot app, backed by and scm, when we change a configuration property, the the spring-boot app needs to be restarted to be able to fetch the new config change so we also need this kind grace restart trigger without having to do a redeployment.

**Comment by gsf** [comment]
@japzio As [suggested by Helm](https://github.com/helm/helm/blob/master/docs/charts_tips_and_tricks.md#automatically-roll-deployments-when-configmaps-or-secrets-change), a checksum of the configmap in the annotations is a good way to handle that case.

**Comment by bholagabbar-mt** [comment]
Has there been any update on this ? We're looking to have this feature too.

**Comment by bgrant0607** [comment]
@bholagabbar-mt What is your use case?

cc @kow3ns @janetkuo

**Comment by bholagabbar-mt** [comment]
@bgrant0607 @kow3ns @janetkuo The usecase for our systems is multifold.

1. Secrets updating - I'm sure you realise that there are many companies such as mine, that have build their own abstractions over kubernetes. We have our own container management system that is orchestrated over kubernetes. So the [helm suggestions(https://github.com/kubernetes/kubernetes/issues/13488#issuecomment-431399754) and others are not applicable. To reload secrets from ConfigMaps in the dev cluster, we have to force kill the pods, resulting in a few seconds of downtime. This should not be the case. This is a real usecase for rolling update.

2. This is a bit complicated, but the overall scope, as someone suggested, is to fix abnormal behaviour. We have 4-5 heavy Java apps running on the Play framework. We are encountering a situation where the memory consumption of our java pods linearly rises and then restarts the pods when the memory limit is reached. This is a documented Java issue with a [stackoverflow issue](https://stackoverflow.com/questions/52963152/kubernetes-pod-reporting-more-memory-usage-than-actual-process-consumption) and [Kubernetes Issue](https://github.com/kubernetes/kubernetes/issues/70179) associated to it. Rolling-restart of all the pods in a 3-4 hour period would reset the memory consumption and allow our apps to function smoothly without spikes. 

Hopefully this was convincing enough and this feature can be taken up by someone for development ?

**Comment by montanaflynn** [comment]
@bholagabbar-mt just change an environment variable and it'll trigger a rolling deploy:

```
kubectl patch deployment mydeployment -p '{"spec":{"template":{"spec":{"containers":[{"name":"mycontainer","env":[{"name":"LAST_MANUAL_RESTART","value":"'$(date +%s)'"}]}]}}}}'
```

**Comment by vdboor** [comment]
Another use-case for this: due to a security issue in containerd I'd like to restart all the pods. https://seclists.org/oss-sec/2019/q1/119 Either the cluster goes completely down, or it does a rolling restart. Having a restart command would make a huge difference!

**Comment by 8427003** [comment]
@realfresh you are the best practices. add `annotation:{creatTime: 12312312} ` in rancher!

**Comment by m-aciek** [comment]
kubectl set env deployment mydeployment --env="LAST_RESTART=$(date)" --namespace ...

seems to be minimal command that does the job. It is example of use of imperative configuration.

**Comment by bgrant0607** [comment]
cc @monopole @apelisse

**Comment by whereisaaron** [comment]
> ~Two~ *Three* and a half years on and people are still crafting new workarounds, with dummy env vars, dummy labels, *dummy annotations*, ConfigMap and Secret watcher sidecars, scaling to zero, and straight out rolling-update shell scripts to simulate the ability the trigger a rolling update. Is this still something cluster admins should not be allowed to do honestly, without the tricks?

Rolling updates are still a pretty popular activity for something with apparently no use case 😄

**Comment by dims** [comment]
long-term-issue (note to self)

**Comment by kow3ns** [comment]
1. I don't see a way to do this without letting imperative logic bleed into a declarative API, thereby, breaking our convention of keeping the APIs declarative and implementing imperative behavior in the controllers, and introducing incompatibilities with most CI/CD practices.
2. I could imagine an RollingRestart API and controller where the creation of a RollingRestart resource caused the controller to restart Pods 1 by 1 via eviction (thus respecting disruption budgets),  but such a controller could be implemented as a CRD (i.e. I see know reason why we have to do this in tree).
3. The declarative approach, e.g. adding a lastRestartedAt=TIMESTAMP annotation doesn't seem like a hack to me.
4. If anyone would like to offer a declarative design and contribution to implement this feature (in tree or otherwise) please consider authoring a KEP PR against the [enhancements repo](https://github.com/kubernetes/enhancements). If a declarative, built in, implementation can be designed I'd be happy to review/shepard in workloads APIs. If a CRD controller like [2] is offered we could sponsor it in SIG Apps as an community sponsored extension.

**Comment by jethrogb** [comment]
> The declarative approach, e.g. adding a lastRestartedAt=TIMESTAMP annotation doesn't seem like a hack to me.

It's not a hack, there should just be a shorthand command line for it.

**Comment by apelisse** [comment]
Someone could build a [krew](https://github.com/kubernetes-sigs/krew) plugin that sends the patch. `kubectl restart-deployment <deployment_name>`?

**Comment by whereisaaron** [comment]
`kubectl rollout restart` that patches a deployment/daemonset/statefulset to trigger a fresh 'rollout'? 

That is in line with @kow3ns's approach (3), and makes some sense since you can then watch/pause/resume the rollout you just started with the other `kubectl rollout` commands.

**Comment by apelisse** [comment]
@whereisaaron I'll see if I can send a patch for that.

**Comment by bgrant0607** [comment]
For rolling out new secrets and configmaps, my recommendation is still #22368: create new ones. That provides a means for both controlling the rollout and also rolling back. We just need to finish GC of old objects:
https://github.com/kubernetes/community/pull/1163
https://github.com/kubernetes/community/pull/2287

Documenting and/or supporting (in kustomize, kubectl, or a kubectl plugin) the recommended way to do a rolling restart with the existing API is reasonable, though.

cc @monopole

**Comment by bgrant0607** [comment]
As for new images, CI/CD or a controller that resolves tags to digest: #1697.

**Comment by bgrant0607** [comment]
Moving unhappy pods was intended to be performed by the descheduler (https://github.com/kubernetes-incubator/descheduler) or something like it, that could consume container status, core metrics, and even custom metrics:

https://github.com/kubernetes/community/blob/master/contributors/design-proposals/scheduling/rescheduler.md
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/scheduling/rescheduling.md

**Comment by apelisse** [comment]
Also, official documentation on how to handle secrets and configmap: https://kubectl.docs.kubernetes.io/pages/app_management/secrets_and_configmaps.html

**Comment by pc-rshetty** [comment]
Rolling restart is very much needed . 1 example is we get all secrets from SSM from AWS. If we change some secret from SSM we would like to do a rolling restart so that the pods now pick up the latest when they boot. Also some times there are application issues which needs rolling restart till actual fix lands on production.

**Comment by apelisse** [comment]
OK, `kubectl rollout restart` has merged!

**Comment by montanaflynn** [comment]
That's great to hear after nearly 4 years, thanks! 

I believe the merged PR only supports deployments, is that correct? 

Some people in this issue have expressed the need for restarting daemonsets and statefulsets as well.

**Comment by e-nikolov** [comment]
@apelisse is there also a way to do this via the sdk or just kubectl?

**Comment by apelisse** [comment]
@e-nikolov What is the SDK?

**Comment by e-nikolov** [comment]
I meant the Go client that can be used for talking to kubernetes from Go programs.

**Comment by apelisse** [comment]
No, you would have to re-implement the (very simple) logic that is implemented in kubectl.

**Comment by kivagant-ba** [comment]
> OK, kubectl rollout restart has merged!

What `kubectl` version will have it?

**Comment by apelisse** [comment]
> What kubectl version will have it?

Kubernetes 1.15

**Comment by dimileeh** [comment]
Our GKE cluster on "rapid" release channel has upgraded itself to Kubernetes 1.16 and now `kubectl rollout restart` has stopped working: 

> kubectl rollout restart deployment myapp
> error: unknown command "restart deployment myapp"

**Comment by apelisse** [comment]
hey @dimileeh 

Do you happen to know what version of kubectl you're using now? and what version you used before? I'd love to know if there was a regression, but at the same time I'd be surprised if the feature had entirely disappeared.

With regard to the GCS thing, and knowing very little about your use-case so sorry if it makes no sense: I would suggest that the gcs model get a different name every time they get modified (maybe suffix with their hash), and that the name would be included in the deployment. Updating the deployment to use the new files would automatically trigger a rollout. This give you the ability to roll-back to a previous deployment, have a better understanding of the changes happening to the models, etc.

**Comment by dimileeh** [comment]
hi @apelisse, thank you for your response!

When I run `kubectl version` from Google Cloud Terminal, I get the following:

```
Client Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.11-dispatcher", GitCommit:"2e298c7e992f83f47af60cf4830b11c7370f6668", GitTreeState:"clean", BuildDate:"2019-09-19T22:20:12Z", GoVersion:"go1.11.13", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.0-gke.20", GitCommit:"d324c1db214acfc1ff3d543767f33feab3f4dcaa", GitTreeState:"clean", BuildDate:"2019-11-26T20:51:21Z", GoVersion:"go1.12.11b4", Compiler:"gc", Platform:"linux/amd64"}
```

When I tried to upgrade kubectl via `gcloud components update`, it said I'm already using the latest versions of all products. Therefore, I think my kubectl version stayed the same while the K8S cluster upgraded from 1.15 to 1.16.

The Kubenetes documentation 1.17, 1.16 and 1.15 has nothing about `kubectl rollout restart` feature. So I wonder if your valuable contribution could have disappeared from 1.16?

-----------------
Thank you for your suggestion on model versioning. We thought about that but then, since we retrain our models every day, we thought we'd start accumulating too many models (and they are quite heavy). Of course, we could use some script to clean up old versions after some time, but so far we've decided to keep it simple relying on ```kubectl rollout restart``` :)

**Comment by apelisse** [comment]
I can see the docs here:
https://v1-16.docs.kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#-em-restart-em-

**Comment by dimileeh** [comment]
Ah, thank you, I was looking here:
https://v1-16.docs.kubernetes.io/docs/reference/kubectl/cheatsheet/

**Comment by apelisse** [comment]
Thank you very much for that link, I’ll make sure it gets updated !

**Comment by apelisse** [comment]
@dimileeh PTAL https://github.com/kubernetes/website/pull/18224 (I'll cherry-pick in relevant branches once this gets merged).

**Comment by apelisse** [comment]
@dimileeh I think I figured out what's wrong with your kubectl version, we'll be working on it.

**Comment by anuragtr** [comment]
Yes, we also have use case of re-starting pod without code change, after updating the configmap. This is to update a ML model without re-deploying the service.

**Comment by montanaflynn** [comment]
@anuragtr with latest versions you can run 

```
kubectl rollout restart deploy NAME
```

**Comment by mauri870** [comment]
I was using a custom command for that [1], glad it is now in the standard kubectl! Thanks

[1] https://github.com/mauri870/kubectl-renew

**Comment by japzio** [comment]
> @anuragtr with latest versions you can run
> 
> ```
> kubectl rollout restart deploy NAME
> ```

@countrogue

**Comment by shoce** [comment]
As i understand now if i have a newer docker image tagged as :latest and a deployment using the image tagged :latest, with the `kubectl rollout restart` and `imagePullPolicy: Always` specified in the container template it is possible to restart the pods and they will pull the newer image.

**BUT is there a way to ask Kubernetes to check if the image has really changed and restart pods ONLY if the image differs from that one used by the pods?**

I am migrating services that are managed by docker-compose now, and currently i run `docker-compose up -d` and it restarts a service (using :latest tagged image) only if another version of the image (still tagged as :latest) is available now.

**Comment by AndrewFarley** [comment]
@shoce I believe you may misunderstand how kubernetes and this latest tag concept works.  

Simply put, if you use a dynamic tag (like latest) that can have multiple different images at any given time, you can never guarantee you are always using the same version of that tag (aka the same image).  Kubernetes doesn’t do a “lookup” like you seem to assume to check what sha sum the current t deployment needs to pull from your container registry.  This all is possible whether or not you have image pull policy set.

An example of it helps, is let’s say you have 3 nodes and a deployment of your “widgets” service that has 3 replicas specified and let’s say your image pull policy is always.  Let’s say you trigger an update to your service (although I do not know how since your image tag didn’t change.  So let’s say you do something silly which I’ve seen before like set the current date into an annotation).  The second this triggers on the first node it will try to  bring up a new pod with your latest latest, but before that gets healthy let’s say your CI system or your dev pushed a new latest.  Then after this your first pod got healthy and a new pod on the second node tried to come up. This one would now be using the newer latest.

TL;DR: Do not use latest tag for anything ever for the most part.  It’s really bad practice.  All registries I know of have a feature you can enable which disallows pushing over an existing tag exactly for this reason.   It’s bad.  There are simple use cases for latest (Eg useful for internal ci images and tooling and can be useful in docker files)  but you should understand when you use and not use them.  Deploying something into kubernetes with a latest tag is generally viewed as “doing something wrong/funny” in my experience.

**Comment by shoce** [comment]
@AndrewFarley thanks, your explanation helped me a lot and the example is something i could not see before. Actually what i do is using :develop and :master tagging of docker images built against the corresponding git branches and deploy them to dev an staging environments as soon as possible. As my dev and staging environments are hosted on a single host i did not worry much.

After reading many discussions on :latest tags and Kubernetes i finally agree to drop using :latest tagging with Kubernetes even for dev and staging environments. It was simple to use with docker-compose but not ok with Kubernetes.

So i have two related questions now and i would appreciate any leads. It might seem off topic but i believe these are the issues blocking people to drop using :latest tags.
1. What would be the best way to trigger deployments' updates? Instead of the cron job i have now that does `docker pull` and `docker-compose up -d`. I have a feeling it is too much coupling to make the build system to trigger updates to the cluster. I prefer the cluster to check for updates available but may be i am wrong here.
2. How to clean old docker images in private docker registry? Now i have :develop and :master tags only and no need to make cleaning. But with proper tagging i will have to keep only the latest :develop-timestamp-or-hash image and delete all older images. This really adds me headache as i try to have minimum possibilities of running out of disk space.

**Comment by AndrewFarley** [comment]
@shoce You should google this problem for your registry provider (eg: "Automatically delete old images on REGISTRY_PROVIDER" or perhaps "Delete untagged images on REGISTRY_PROVIDER")

Each registry tends to have their own API and/or tools to handle this.  AWS, for example, has a built-in [Lifecycle Policy](https://docs.aws.amazon.com/AmazonECR/latest/userguide/LifecyclePolicies.html) which you can configure to automatically delete images without requiring much effort on your part.  In the past on "simpler" registries I've been known to write simple scripts to query the old images and delete them.

Good luck!
